src/main/java/usi/si/seart/repository/criteria/KeyValueCriteria.java

```diff
@@ -4,13 +4,63 @@
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import lombok.experimental.FieldDefaults;
+import usi.si.seart.model.GitRepo;
 import usi.si.seart.repository.operation.BinaryOperation;
 
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Path;
+import javax.persistence.criteria.Predicate;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
 @Getter
 @AllArgsConstructor
 @FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
 public class KeyValueCriteria implements Criteria {
     String key;
     Object value;
     BinaryOperation operation;
```

Comment:
I have a suggestion for the class signature: Make both `KeyValueCriteria` and `NestedKeyValueCriteria` generic. As such the highlighted lines would change to:

```java
public class KeyValueCriteria<T> implements Criteria {
    String key;
    T value;
    BinaryOperation operation;
    // ...
}
```

By switching to a generic type, we can also type the `Path` object when getting a field:
```java
path.<T>get(key)
```

The JPA documentation [also recommends this](https://docs.oracle.com/javaee/6/api/javax/persistence/criteria/Path.html#get(java.lang.String)). As a matter of fact, doing this will also completely eliminate the need for performing the awful `instanceof` checks that I was doing before.

Reply:
Actually, the only exceptions to the last sentence: `LIKE` only works with strings and `IN` should match only `Collection<T>`

Reply:
I've just had another epiphany regarding generification. You can go a step further than what I suggested, and refactor the `KeyValueCriteria` and `NestedKeyValueCriteria` to use the `SingularAttribute` and `PluralAttribute` instead of the `String` for the key. So for example, the first of the two mentioned classes would become:
```java
public class KeyValueCriteria<E, T> implements Criteria {
    SingularAttribute<E, T> key;
    T value;
    BinaryOperation operation;
    // ...
}
```
And when instantiating the class, you would for instance do:
```java
new KeyValueCriteria<GitRepo, Long>(GitRepo_.contributors, contributors.lowerEndpoint(), BinaryOperation.GREATER_THAN_EQUAL);
```
Note the usage of `GitRepo_.contributors` instead of `GitRepo_.CONTRIBUTORS`.
I recommend you first generify the value, then also try changing the key type.